File: CP/JanuaryCookOff2018/prob3.cpp
#include <bits/stdc++.h>
using namespace std;
int A[100009];
int D1[100009];
int D2[100009];
bool ok(int i)
{
	if (i % 2 == 1)
		return A[i] < A[i + 1];
	else
		return A[i] > A[i + 1];
}
int N;
int ans()
{
	int i = (N + 1) / 2;
	int j = (N + 2) / 2;
	int id = 0;
	D1[id] = N + 2;
	D2[id] = N + 2;
	if (i == j)
	{
		D1[id] = 0;
		D2[id] = 0;
	}
	else
	{
		if (ok(i))
			D1[id] = 0;
		swap(A[i], A[j]);
		if (ok(i))
			D2[id] = 1;
		swap(A[i], A[j]);
	}
	while (true)
	{
		i--;
		j++;
		if (i < 1)
			break;
		id++;
		D1[id] = N + 2;
		D2[id] = N + 2;
		{
			if (ok(i) && ok(j - 1))
				D1[id] = min(D1[id], D1[id - 1]);
			swap(A[i + 1], A[j - 1]);
			if (ok(i) && ok(j - 1))
				D1[id] = min(D1[id], D2[id - 1]);
			swap(A[i + 1], A[j - 1]);
		}
		swap(A[i], A[j]);
		{
			if (ok(i) && ok(j - 1))
				D2[id] = min(D2[id], 1 + D1[id - 1]);
			swap(A[i + 1], A[j - 1]);
			if (ok(i) && ok(j - 1))
				D2[id] = min(D2[id], 1 + D2[id - 1]);
			swap(A[i + 1], A[j - 1]);
		}
		swap(A[i], A[j]);
	}
	return min(D1[id], D2[id]);
}
int main()
{
	int T;
	cin >> T;
	while (T--)
	{
		cin >> N;
		for (int i = 1; i <= N; i++)
			cin >> A[i];
		int out = N + 2;
		out = min(out, ans());
		for (int i = 1; i <= N; i++)
			A[i] *= -1;
		out = min(out, ans());
		if (out > N)
			out = -1;
		cout << out << '\n';
	}
}

File: CP/JanuaryCookOff2018/prob3.py
'''input
1
6
1 2 3 4 5 6

'''
from bisect import bisect_left as bl
from itertools import combinations as c
from collections import defaultdict as df
from math import factorial as F
from math import ceil


for _ in range(input()):
	n=input()
	l=[int(x) for x in raw_input().split()]

	fix=[0]*n
	if n==1 :
		print 0
	elif n==2 and len(set(l))==1:
		print -1
	elif n==2:
		print 0



	elif n&1:
		dp=[]
		for i in range(n):
			dp.append([0,0])


		def satis(ind):
			if ind&1==0:
				return l[ind]>l[ind-1] and l[n-ind-1]>l[n-ind]
			else:
				return l[ind]<l[ind-1] and l[n-ind-1]<l[n-ind]

		for ind in range(n/2+1,n):
			if satis(ind):
				dp[ind][0]=dp[ind-1][0]
			else:
				dp[ind][0]=9999999

			l[ind],l[n-ind-1]=l[n-ind-1],l[ind]

			if satis(ind):
				dp[ind][1]=dp[ind-1][0]+1
			else:
				dp[ind][1]=9999999

			l[ind-1],l[n-ind]=l[n-ind],l[ind-1]

			if satis(ind):
				dp[ind][1]=min(dp[ind][1],dp[ind-1][1]+1)
			else:
				dp[ind][1]=min(dp[ind][1],9999999)

			l[ind],l[n-ind-1]=l[n-ind-1],l[ind]

			if satis(ind):
				dp[ind][0]=min(dp[ind-1][1],dp[ind][0])
			else:
				dp[ind][0]=min(dp[ind][0],9999999)

			l[ind-1],l[n-ind]=l[n-ind],l[ind-1]
		a1=min(dp[-1])
		dp=[]
		for i in range(n):
			dp.append([0,0])


		def satis(ind):
			if ind&1:
				return l[ind]>l[ind-1] and l[n-ind-1]>l[n-ind]
			else:
				return l[ind]<l[ind-1] and l[n-ind-1]<l[n-ind]

		for ind in range(n/2+1,n):
			if satis(ind):
				dp[ind][0]=dp[ind-1][0]
			else:
				dp[ind][0]=9999999

			l[ind],l[n-ind-1]=l[n-ind-1],l[ind]

			if satis(ind):
				dp[ind][1]=dp[ind-1][0]+1
			else:
				dp[ind][1]=9999999

			l[ind-1],l[n-ind]=l[n-ind],l[ind-1]

			if satis(ind):
				dp[ind][1]=min(dp[ind][1],dp[ind-1][1]+1)
			else:
				dp[ind][1]=min(dp[ind][1],9999999)

			l[ind],l[n-ind-1]=l[n-ind-1],l[ind]

			if satis(ind):
				dp[ind][0]=min(dp[ind-1][1],dp[ind][0])
			else:
				dp[ind][0]=min(dp[ind][0],9999999)

			l[ind-1],l[n-ind]=l[n-ind],l[ind-1]
		a2=min(dp[-1])
		ans=min(a1,a2)
		if ans>999999:
			print -1
		else:
			print ans




	else:


		def satis(ind):
			if ind&1==0:
				return l[ind]>l[ind-1] and l[n-ind-1]<l[n-ind]
			else:
				return l[ind]<l[ind-1] and l[n-ind-1]>l[n-ind]
		dp=[]



		for i in range(n):

			dp.append([0,1])

		for ind in range(n/2+1,n):
			if satis(ind):
				dp[ind][0]=dp[ind-1][0]
			else:
				dp[ind][0]=9999999

			l[ind],l[n-ind-1]=l[n-ind-1],l[ind]

			if satis(ind):
				dp[ind][1]=dp[ind-1][0]+1
			else:
				dp[ind][1]=9999999

			l[ind-1],l[n-ind]=l[n-ind],l[ind-1]

			if satis(ind):
				dp[ind][1]=min(dp[ind][1],dp[ind-1][1]+1)
			else:
				dp[ind][1]=min(dp[ind][1],9999999)

			l[ind],l[n-ind-1]=l[n-ind-1],l[ind]

			if satis(ind):
				dp[ind][0]=min(dp[ind-1][1],dp[ind][0])
			else:
				dp[ind][0]=min(dp[ind][0],9999999)

			l[ind-1],l[n-ind]=l[n-ind],l[ind-1]
		a1=min(dp[-1])


		#print dp

		def satis(ind):
			if ind&1:
				return l[ind]>l[ind-1] and l[n-ind-1]<l[n-ind]
			else:
				return l[ind]<l[ind-1] and l[n-ind-1]>l[n-ind]

		dp=[]
		for i in range(n):
			dp.append([0,1])

		for ind in range(n/2+1,n):
			if satis(ind):
				dp[ind][0]=dp[ind-1][0]
			else:
				dp[ind][0]=9999999

			l[ind],l[n-ind-1]=l[n-ind-1],l[ind]

			if satis(ind):
				dp[ind][1]=dp[ind-1][0]+1
			else:
				dp[ind][1]=9999999

			l[ind-1],l[n-ind]=l[n-ind],l[ind-1]

			if satis(ind):
				dp[ind][1]=min(dp[ind][1],dp[ind-1][1]+1)
			else:
				dp[ind][1]=min(dp[ind][1],9999999)

			l[ind],l[n-ind-1]=l[n-ind-1],l[ind]

			if satis(ind):
				dp[ind][0]=min(dp[ind-1][1],dp[ind][0])
			else:
				dp[ind][0]=min(dp[ind][0],9999999)

			l[ind-1],l[n-ind]=l[n-ind],l[ind-1]
		a2=min(dp[-1])
		#print dp
		ans=min(a1,a2)
		#print satis(3)
		if ans>999999:
			print -1
		else:
			print ans
